local_settings.sh.in

```diff
@@ -16,45 +16,98 @@ MODULES=${MODULES:-@MODULES@}
+["${ROOTDIR}tftpboot(/.*)?"]='tftpdir_t'
+["$LOCALSTATEDIR/log/pacemaker.log.*"]='cluster_var_log_t'
+["$LOCALSTATEDIR/log/pacemaker(/.*)?"]='cluster_var_log_t'
+)
+
+# This associative array ("dict", "map", "hash") will holds all the current
+# custom fcontext settings
+declare -A existing_custom_fcontext
+
+# This associative array ("dict", "map", "hash") will holds all the differences
+# we have between existing custom fcontext, and what we really want.
+# It will be used in different methods in this script.
+declare -A diff_fcontext
+
+# Let's populate the existing_custom_fcontext
+for l in $(semanage fcontext -Cln | awk -F '[: ]+' '{print $1";"$6}'); do
+	existing_custom_fcontext+=([$(echo $l|cut -d ';' -f1)]=$(echo $l|cut -d ';' -f2))
+done
+
+# Let's populate the diff_fcontext
+for i in "${!custom_fcontext[@]}"; do
+  if [[ ! -v existing_custom_fcontext["$i"] ]]; then
+    diff_fcontext+=([$i]=${custom_fcontext[$i]})
+  elif [[ ${existing_custom_fcontext["$i"]} != "${custom_fcontext[$i]}" ]]; then
+    diff_fcontext+=([$i]=${custom_fcontext[$i]})
+  fi
+done
+
 do_echo() {
 	if [ $QUIET -eq 0 ]; then
 		return
 	fi
 	echo $*
 }
 
+function get_clean_name() {
+  if [[ "${1}" =~ '(' ]]; then
+    echo "$1" | cut -d '(' -f1
+  else
+    echo $1
+  fi
+}
 
 relabel_files()
 {
-	local opts=""
-
-	do_echo "Relabeling files..."
-	if [ $QUIET -ne 0 ]; then
-		opts="-v"
-	fi
-
-	# Setfiles is a lot like restorecon, except it takes a policy
-	# on-disk instead of looking at the kernel-loaded policy
-	# So, it works inside of image builds.
-	#
-	# At a minimum, we need a line for each entry we have where we are
-	# custominzing the label using 'semanage' below, but also
-	# others, when applicable.
-	$SBINDIR/setfiles $opts -F $ROOTDIR/etc/selinux/targeted/contexts/files/file_contexts \
-		$BINDIR/swift* \
-		$LOCALSTATEDIR/run/swift \
-		$LOCALSTATEDIR/swift \
-		$SHAREDSTATEDIR/nova/.ssh \
-		$SHAREDSTATEDIR/designate/bind9 \
-		$SHAREDSTATEDIR/vhost_sockets \
-		/srv \
-		$BINDIR/neutron* \
-		$BINDIR/swift-object-* \
-		${ROOTDIR}httpboot \
-		${ROOTDIR}tftpboot \
-		$LOCALSTATEDIR/run/redis \
-		$LOCALSTATEDIR/log \
-	2> /dev/null || :
+  local opts=""
+
+  do_echo "Relabeling files..."
+  if [ $QUIET -ne 0 ]; then
+    opts="-v"
+  fi
+
+  # Setfiles is a lot like restorecon, except it takes a policy
+  # on-disk instead of looking at the kernel-loaded policy
+  # So, it works inside of image builds.
+  #
+  # At a minimum, we need a line for each entry we have where we are
+  # customizing the label using 'semanage' below, but also
+  # others, when applicable.
+  declare -a f_list
+  for i in "${!custom_fcontext[@]}"; do
+    f_list=("${f_list[@]}" "$(get_clean_name "$i")")
+  done
+  $SBINDIR/setfiles $opts -F $ROOTDIR/etc/selinux/targeted/contexts/files/file_contexts \
+    $LOCALSTATEDIR/run/swift \
+    $LOCALSTATEDIR/swift \
+    /srv \
+    $LOCALSTATEDIR/run/redis \
+    $LOCALSTATEDIR/log \
+    ${!f_list[@]} 2> /dev/null || :
```

Comment:
I'm still running tests, but this command is not working for me so far. (I'm only testing on RHEL8.2 for now.) If I echo it, I get the following result:

/sbin/setfiles -v -F //etc/selinux/targeted/contexts/files/file_contexts     /var/run/swift     /var/swift     /srv     /var/run/redis     /var/log     0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 2> /dev/null || :

Poking around, the custom_fcontext content looks correct, and get_clean_name returns the expected output as well. So I think this might come from the way the f_list is built on line 102? I'm still refreshing my bash hieroglyph understanding so the reason isn't immediately obvious to me!

Reply:
It looks like that might actually be due to the exclamation mark on line 110? This is causing the keys in the array to show instead of the values. Do we even need an associate array here? It seems like a simple list would work as nicely and may reduce some of the complexity? 

I don't mind switch from tabs to spaces for indentation, but could you make these 4 spaces? I think that would help with readability.

Even with line 110 changed to use the values instead of keys, I'm seeing issue with getting new file contexts to actually apply. Not sure yet if it's due to my environment, or something else is going on.